### PR TITLE
RFC: Updated Contributing.md

### PR DESCRIPTION
**Pull request by ambrosejcarr**
What's still missing from this doc?

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -2,18 +2,49 @@
...
+2. Create a new python class `NewFilter(PipelineComponent)` in the `new_filter.py` class.
+
+3. Implement all required methods from the base class.
+
+4. For the command line arguments to mimic the API, the arguments passed to the CLI must exactly match the names of the parameters for `NewFilter.__init__`, after dashes are automatically converted by argparse.
+For example, `--foo-bar` would convert to `foo_bar` and init must accept such an argument: `NewFilter.__init__(foo_bar, ..., **kwargs)`
+
+5. `NewFilter.__init__()` must have a `**kwargs` parameter to accept arbitrary CLI args.
+
+That's it! your pipeline component will automatically register and be available under `starfish filter` in the CLI.
```
Comment (ttung) [inline CONTRIBUTING.md]: I don't think this is true.  What you're describing is the outcome of a new FilterAlgorithm, not a new PipelineComponent.  Wiring up a new PipelineComponent is a bit more complicated...

Comment (ambrosejcarr) [reply CONTRIBUTING.md]: Ah true, I'll clean up this language. 

I _am_ describing a new Algorithm, and I made that decision deliberately because (1) I don't think we've missed many obvious pipeline components (2) I hope a user would probably contact us before writing a new pipeline component and (3) not describing how to wire up a new component will bias a user towards working in the existing workflow. 

Does this logic make sense to you?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -2,18 +2,49 @@
...
-# Code of conduct
+# How can I contribute?
+Starfish is designed to specify pipeline recipes for image processing. To support this use, the library is composed as a series of `pipeline_component` modules.
+The objects that sub-class `pipeline_component` spawn a command-line interface that should identically track the API of the python library.
 
+A typical starfish run consists of running one or more image processing filter stages, and identifying features through either a spot- or pixel-based approach.
+The identified features are then decoded into the genes that they correspond to by mapping the fluorescence channel (and optionally hybridization round) using a codebook.
+Finally, the filtered data are segmented, identifying which cell each feature belongs to.
+
+Implementing a new module of an existing `pipeline_component` is a simple process. For example, to add another image filtering step, one would:
```
Comment (ttung) [inline CONTRIBUTING.md]: I see that the terminology of "module" came from the spec.  I would argue that we should ditch the term "module" as it is ridiculously overloaded.

I would just rephrase this as "Creating a new algorithm for an existing \`pipeline_component\` ..."

Comment (ttung) [reply CONTRIBUTING.md]: Also, "For example, to add a new image filter, one would:"

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -2,18 +2,49 @@
...
+
+1. Create a new python file `new_filter.py` in the `starfish/pipeline/filter/` directory.
+
+2. Create a new python class `NewFilter(PipelineComponent)` in the `new_filter.py` class.
```
Comment (ttung) [inline CONTRIBUTING.md]: This, I believe, is incorrect.  You would want to subclass `FilterAlgorithmBase` or something akin to that.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -2,18 +2,49 @@
...
+
+3. Implement all required methods from the base class.
+
+4. For the command line arguments to mimic the API, the arguments passed to the CLI must exactly match the names of the parameters for `NewFilter.__init__`, after dashes are automatically converted by argparse.
+For example, `--foo-bar` would convert to `foo_bar` and init must accept such an argument: `NewFilter.__init__(foo_bar, ..., **kwargs)`
+
+5. `NewFilter.__init__()` must have a `**kwargs` parameter to accept arbitrary CLI args.
+
```
Comment (ttung) [inline CONTRIBUTING.md]: starfish/pipeline/filter/__init__.py needs to import the new filter as well.

Comment (ambrosejcarr) [reply CONTRIBUTING.md]: Ah, good catch.
